animations-dbus/animations-dbus-server-object.c

```diff
@@ -735,3 +744,70 @@ animations_dbus_server_new_with_connection_async (AnimationsDbusServerEffectFact
                               "effect-factory", factory,
                               NULL);
 }
+
+/**
+ * animations_dbus_server_stop:
+ * @self: an #AnimationsDbusServer
+ * @cancellable: (nullable): a #GCancellable, or %NULL
+ * @error: return location for a #GError, or %NULL
+ *
+ * Stop the server, unexport all D-Bus objects, and unown the well-known
+ * D-Bus name for the service.
+ *
+ * Once stopped, the server object can only be unreffed and destroyed: it cannot
+ * be restarted.
+ *
+ * Returns: %TRUE on success, %FALSE otherwise
+ */
+gboolean
+animations_dbus_server_stop (AnimationsDbusServer  *self,
+                             GCancellable          *cancellable,
+                             GError               **error)
+{
+  AnimationsDbusServerPrivate *priv = animations_dbus_server_get_instance_private (self);
+  gboolean error_seen = FALSE;
+
+  g_return_val_if_fail (ANIMATIONS_DBUS_IS_SERVER (self), FALSE);
+  g_return_val_if_fail (cancellable == NULL || G_IS_CANCELLABLE (cancellable), FALSE);
+  g_return_val_if_fail (error == NULL || *error == NULL, FALSE);
+
+  while (priv->animatable_surfaces != NULL && priv->animatable_surfaces->len > 0)
+    {
+      AnimationsDbusServerSurface *surface = g_ptr_array_index (priv->animatable_surfaces, 0);
+
+      if (!animations_dbus_server_unregister_surface (self, surface, NULL))
+        error_seen = TRUE;
+    }
+
+
+  if (priv->animation_managers != NULL)
+    {
+      GHashTableIter iter;
+      gpointer key;
+
+      g_hash_table_iter_init (&iter, priv->animation_managers);
+
+      while (g_hash_table_iter_next (&iter, &key, NULL))
+        {
+          const gchar *name = key;
+
+          unregister_client (self, name);
```

Comment:
Sorry, missed this in the prev review - wouldn't that also give us issues with removing while iterating? The doc for `g_hash_table_iter_init` says:
```
Initializes a key/value pair iterator and associates it with hash_table . Modifying the hash table after calling this function invalidates the returned iterator.
```

Reply:
Fixed